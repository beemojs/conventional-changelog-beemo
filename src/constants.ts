import { Group } from './types';

export const COMMIT_FORMAT_PREFIX =
	// eslint-disable-next-line unicorn/no-unsafe-regex
	/^(break|breaking|build|chore|ci|cd|deps|docs|feature|fix|internal|misc|new|patch|perf|release|revert|security|style|styles|test|tests|type|types|update)(?:\(([a-zA-Z0-9\-., ]+)\))?:/u;

export const GROUPS: Group[] = [
	{
		bump: 'major',
		emoji: '💥',
		label: 'Breaking',
		types: ['break', 'breaking'],
	},
	{
		bump: 'patch',
		emoji: '📦',
		label: 'Dependencies',
		types: ['deps'],
	},
	{
		emoji: '📘',
		label: 'Docs',
		types: ['docs'],
	},
	{
		bump: 'patch',
		emoji: '🐞',
		label: 'Fixes',
		types: ['fix', 'patch'],
	},
	{
		emoji: '🛠',
		label: 'Internals',
		types: ['ci', 'cd', 'build', 'test', 'tests', 'internal'],
	},
	{
		bump: 'patch',
		emoji: '📋',
		label: 'Misc',
		types: ['chore', 'misc'],
	},
	{
		bump: 'major',
		emoji: '🎉',
		label: 'Release',
		types: ['release'],
	},
	{
		bump: 'patch',
		emoji: '↩️',
		label: 'Reverts',
		types: ['revert'],
	},
	{
		bump: 'patch',
		emoji: '🔑',
		label: 'Security',
		types: ['security'],
	},
	{
		bump: 'patch',
		emoji: '🎨',
		label: 'Styles',
		types: ['style', 'styles'],
	},
	{
		bump: 'patch',
		emoji: '⚡️',
		label: 'Performance',
		types: ['perf'],
	},
	{
		bump: 'patch',
		emoji: '⚙️',
		label: 'Types',
		types: ['type', 'types'],
	},
	{
		bump: 'minor',
		emoji: '🚀',
		label: 'Updates',
		types: ['new', 'update', 'feature'],
	},
];
