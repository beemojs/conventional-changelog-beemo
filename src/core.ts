/**
 * @copyright   2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

export { checkCommitFormat } from './checkCommitFormat';
export { getTypeGroup } from './getTypeGroup';
export { parserOpts } from './parserOpts';
export { whatBump } from './whatBump';
export { writerOpts } from './writerOpts';
export * from './constants';
export * from './types';
