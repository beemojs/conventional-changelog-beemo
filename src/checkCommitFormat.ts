import { COMMIT_FORMAT_PREFIX } from './constants';
import type { CommitType } from './types';

export function checkCommitFormat(
	commit: string,
): { breaking: boolean; scope: string; type: CommitType } | null {
	const match = commit.match(COMMIT_FORMAT_PREFIX);

	if (!match) {
		return null;
	}

	const type = match[1] as CommitType;

	return {
		breaking: !!match[3] || type === 'breaking' || type === 'break',
		scope: match[2] || '',
		type,
	};
}
