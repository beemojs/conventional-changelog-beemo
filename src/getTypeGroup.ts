import { GROUPS } from './constants';
import type { CommitType, Group } from './types';

export function getTypeGroup(type: CommitType): Group {
	const group = GROUPS.find((g) => g.types.includes(type));

	if (!group) {
		throw new Error(`Cannot find group for type "${type}".`);
	}

	return group;
}
