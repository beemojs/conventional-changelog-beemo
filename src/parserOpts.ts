import { COMMIT_FORMAT_PREFIX } from './constants';
import { ParserOptions } from './types';

export const parserOpts: Partial<ParserOptions> = {
	headerCorrespondence: ['type', 'scope', 'message'],
	// Keep in sync with checkCommitFormat
	headerPattern: new RegExp(`^${COMMIT_FORMAT_PREFIX.source} (.*)$`, 'u'),
	mergeCorrespondence: ['pr', 'source'],
	mergePattern: /^Merged? pull request #(\d+) from (.*)/u,
	noteKeywords: 'Note',
	revertCorrespondence: [],
	revertPattern: /^Revert/u,
};
