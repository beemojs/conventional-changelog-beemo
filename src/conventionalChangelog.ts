import { parserOpts } from './parserOpts';
import { writerOpts } from './writerOpts';

export const conventionalChangelog = { parserOpts, writerOpts };
