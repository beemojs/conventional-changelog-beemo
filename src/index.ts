/**
 * @copyright   2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import { parserOpts } from './parserOpts';
import { writerOpts } from './writerOpts';
import { whatBump } from './whatBump';

export * from './types';

export default function createPreset() {
	return {
		parser: parserOpts,
		writer: writerOpts,
		whatBump,
	};
}
