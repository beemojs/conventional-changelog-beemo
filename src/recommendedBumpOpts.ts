import getTypeGroup from './getTypeGroup';
import parserOpts from './parserOpts';
import { BumpOptions, Group, SemverLevel } from './types';

const options: BumpOptions = {
  parserOpts,

  whatBump(commits) {
    let level: SemverLevel = null;
    let breakings = 0;
    let features = 0;

    commits.forEach((commit) => {
      let group: Group;

      try {
        group = getTypeGroup(commit.type);
      } catch {
        return;
      }

      if (group.bump === 'major') {
        breakings += 1;
        level = 0;
      } else if (group.bump === 'minor') {
        features += 1;
        if (level === null || level === 2) {
          level = 1;
        }
      } else if (group.bump === 'patch') {
        if (level === null) {
          level = 2;
        }
      }
    });

    return {
      level,
      reason: `There are ${breakings} breaking changes and ${features} new features`,
    };
  },
};

export default options;
