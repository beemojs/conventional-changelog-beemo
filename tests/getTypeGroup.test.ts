import { describe, it, expect } from 'vitest';
import { getTypeGroup } from '../src/getTypeGroup';

describe('getTypeGroup()', () => {
	it('returns the group by type', () => {
		expect(getTypeGroup('new')).toEqual({
			bump: 'minor',
			emoji: 'ðŸš€',
			label: 'Updates',
			types: ['new', 'update', 'feature'],
		});
	});

	it('errors for invalid type', () => {
		expect(() => {
			// @ts-expect-error Invalid type
			getTypeGroup('unknown');
		}).toThrowErrorMatchingSnapshot();
	});
});
