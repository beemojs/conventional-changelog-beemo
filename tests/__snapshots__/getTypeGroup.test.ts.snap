// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getTypeGroup() > errors for invalid type 1`] = `[Error: Cannot find group for type "unknown".]`;

exports[`getTypeGroup() errors for invalid type 1`] = `"Cannot find group for type \\"unknown\\"."`;
